File: typescript/packages/core/tools/web/web.ts
#!/usr/bin/env node
/**
 * Web Tool - Fetch web content from URLs with flexible HTTP methods
 * Function-type tool: Exports default async function
 */

import axios, { AxiosRequestConfig, AxiosError } from 'axios';
import { MatimoError, ErrorCode } from '../../src/errors/matimo-error';

interface WebParams {
  url: string;
  method?: string;
  headers?: Record<string, string>;
  body?: string;
  timeout?: number;
  followRedirects?: boolean;
  parseJson?: boolean;
  maxSize?: number;
}

interface WebResult {
  success: boolean;
  url: string;
  statusCode: number;
  statusText: string;
  contentType: string;
  content: unknown;
  headers: Record<string, unknown>;
  size: number;
  duration: number;
  redirects?: string[];
}

/**
 * Fetch URL and return response with optional JSON parsing
 */
export default async function webTool(params: WebParams): Promise<WebResult> {
  const {
    url,
    method = 'GET',
    headers,
    body,
    timeout = 30000,
    followRedirects = true,
    parseJson = true,
    maxSize = 10485760,
  } = params;

  const startTime = Date.now();

  // Validate required parameter
  if (!url) {
    throw new MatimoError('Missing required parameter', ErrorCode.INVALID_PARAMETER, {
      reason: 'url is required',
    });
  }

  // Validate URL
  try {
    new URL(url);
  } catch {
    throw new MatimoError('Invalid URL', ErrorCode.INVALID_PARAMETER, {
      url,
      reason: 'URL must be valid http or https',
    });
  }

  const requestConfig: AxiosRequestConfig = {
    method: method.toUpperCase() || 'GET',
    url,
    timeout,
    maxContentLength: maxSize,
    maxBodyLength: maxSize,
    headers: {
      'User-Agent': 'Matimo/1.0 (AI Agent Tool SDK)',
      Accept: 'application/json, text/plain, text/html, */*',
      ...headers,
    },
    validateStatus: () => true, // Don't throw on any status code
  };

  if (body && ['POST', 'PUT', 'PATCH'].includes(method.toUpperCase())) {
    requestConfig.data = body;
  }

  if (!followRedirects) {
    requestConfig.maxRedirects = 0;
  }

  try {
    const response = await axios.request(requestConfig);
    let content: unknown = response.data;

    // Try to parse JSON if requested and content-type suggests JSON
    if (parseJson && response.headers['content-type']?.includes('application/json')) {
      try {
        if (typeof response.data === 'string') {
          content = JSON.parse(response.data);
        } else {
          content = response.data;
        }
      } catch {
        // If JSON parsing fails, keep original content
        content = response.data;
      }
    }

    const result: WebResult = {
      success: response.status >= 200 && response.status < 300,
      url,
      statusCode: response.status,
      statusText: response.statusText || '',
      contentType: (response.headers['content-type'] as string) || 'unknown',
      content,
      headers: response.headers,
      size: JSON.stringify(response.data).length,
      duration: Date.now() - startTime,
    };

    return result;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw new MatimoError(
      'HTTP request failed',
      ErrorCode.NETWORK_ERROR,
      {
        url,
        statusCode: axiosError.response?.status,
        statusText: axiosError.response?.statusText,
        originalError: axiosError.message,
      }
    );
  }
}

File: python/packages/core/src/matimo/tools/web/web.py
"""Web tool — fetch web content from URLs."""
from __future__ import annotations

import time

import httpx


async def run(params: dict) -> dict:  # type: ignore[type-arg]
    url: str = params["url"]
    method: str = params.get("method", "GET").upper()
    headers: dict = params.get("headers", {})  # type: ignore[type-arg]
    body: str | None = params.get("body")
    timeout_ms: int = int(params.get("timeout", 30000))
    parse_json: bool = bool(params.get("parseJson", True))
    follow_redirects: bool = bool(params.get("followRedirects", True))

    merged_headers = {
        "User-Agent": "Matimo/1.0 (AI Agent Tool SDK)",
        "Accept": "application/json, text/plain, text/html, */*",
        **headers,
    }

    start = time.monotonic()
    try:
        async with httpx.AsyncClient(
            follow_redirects=follow_redirects,
            timeout=timeout_ms / 1000.0,
        ) as client:
            req_kwargs: dict = {"headers": merged_headers}  # type: ignore[type-arg]
            if body and method in ("POST", "PUT", "PATCH"):
                req_kwargs["content"] = body
            response = await client.request(method, url, **req_kwargs)

        duration = int((time.monotonic() - start) * 1000)
        content_type = response.headers.get("content-type", "")

        if parse_json and "application/json" in content_type:
            try:
                content = response.json()
            except Exception:
                content = response.text
        else:
            content = response.text

        return {
            "success": response.status_code < 400,
            "url": str(response.url),
            "statusCode": response.status_code,
            "statusText": str(response.status_code),
            "contentType": content_type,
            "content": content,
            "headers": dict(response.headers),
            "size": len(response.content),
            "duration": duration,
        }
    except httpx.TimeoutException:
        duration = int((time.monotonic() - start) * 1000)
        return {
            "success": False,
            "url": url,
            "statusCode": 0,
            "statusText": "Timeout",
            "contentType": "",
            "content": f"Request timed out after {timeout_ms}ms",
            "headers": {},
            "size": 0,
            "duration": duration,
        }
    except Exception as exc:
        duration = int((time.monotonic() - start) * 1000)
        return {
            "success": False,
            "url": url,
            "statusCode": 0,
            "statusText": "Error",
            "contentType": "",
            "content": str(exc),
            "headers": {},
            "size": 0,
            "duration": duration,
        }
